### PR TITLE
Simplify file watching

### PR DESCRIPTION
**Pull request by pranavkm**
* Disable FileSystemWatcher when polling is enabled.
* Let the PollingFileWatcher determine when polling change tokens need to refresh

### CODE REVIEW

File: src/FS.Physical/PhysicalFilesWatcher.cs [hunk 1]
```diff
@@ -21,12 +21,10 @@ namespace Microsoft.Extensions.FileProviders.Physical
     ///     Triggers events on <see cref="IChangeToken" /> when files are created, change, renamed, or deleted.
     ///     </para>
     /// </summary>
-    public class PhysicalFilesWatcher : IDisposable
+    internal sealed class PhysicalFilesWatcher : IFileWatcher
```
Comment (pranavkm) [inline src/FS.Physical/PhysicalFilesWatcher.cs]: Breaking changes for 3.0

File: src/FS.Physical/PhysicalFileProvider.cs [hunk 2]
```diff
@@ -101,37 +100,7 @@ public bool UsePollingFileWatcher
...
-        /// </para>
-        /// <para>
-        /// This property is only effective when <see cref="UsePollingFileWatcher"/> is set.
-        /// </para>
-        /// </summary>
-        /// <value>
-        /// The default value of this property is determined by the value of environment variable named <c>DOTNET_USE_POLLING_FILE_WATCHER</c>.
-        /// When <c>true</c> or <c>1</c>, this property defaults to <c>true</c>; otherwise false.
-        /// </value>
-        public bool UseActivePolling
```
Comment (pranavkm) [inline src/FS.Physical/PhysicalFileProvider.cs]: I took these options out - we'll hopefully have ample time to sort bugs with the polling and Timer our before RTM. Supporting multiple modes just makes the code harder to work with.

File: src/FS.Physical/PhysicalFilesWatcher.cs (hunk 1 above)
Comment (natemcmaster) [reply src/FS.Physical/PhysicalFilesWatcher.cs]: Why?

**Review by natemcmaster — CHANGES_REQUESTED** [review]
I'd like to see if we can make improvements without API breaking changes of this magnitude.

File: src/FS.Physical/PhysicalFilesWatcher.cs (hunk 1 above)
Comment (davidfowl) [reply src/FS.Physical/PhysicalFilesWatcher.cs]: Yea, we shouldn't do this.

Comment (pranavkm) [reply src/FS.Physical/PhysicalFilesWatcher.cs]: Ok. I assumed these were all internal implementation details that we should haven't made public and 3.0 would be a fair opportunity to change it. That said, we would have to change it's behavior \ API. Currently this type supports polling (it's a ctor parameter).

**Review by pranavkm — COMMENTED** [review]
🆙 📅

File: src/FS.Physical/PollingFileWatcher.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        private readonly Timer _timer;
+
+        public PollingFileWatcher(string root, TimeSpan pollingInterval)
+        {
+            _root = root;
+            PollingChangeTokens = new ConcurrentDictionary<string, IPollingChangeToken>(StringComparer.Ordinal);
+
+            // PhysicalFileProvider lazy initializes IFileWatcher instances. It's OK for us to initialize this in the ctor since we anticipate
+            // GetFileChangeToken to be invoked right after.
+            _timer = new Timer(RaiseChangeEvents, state: PollingChangeTokens, dueTime: pollingInterval, period: pollingInterval);
```
Comment (benaadams) [inline src/FS.Physical/PollingFileWatcher.cs]: Don't capture asynclocals onto timer https://github.com/dotnet/corefx/issues/26523#issuecomment-417926130

Comment (pranavkm) [reply src/FS.Physical/PollingFileWatcher.cs]: Ok. I port the workaround to 2.2 too

**Comment by pranavkm** [comment]
🆙 📅

**Comment by mkArtakMSFT** [comment]
Thank you for your effort. As part of the effort to reduce the number of repositories we're going to move the code of this repo to https://github.com/aspnet/aspnetcore repo. Feel free to resubmit a new PR in the target repo as we're closing this PR now.

**Comment by natemcmaster** [comment]
*https://github.com/aspnet/Extensions, not aspnet/AspNetCore.

**Pull request closed without merging**